### PR TITLE
Attribute mismatch in combination with RUC on type

### LINKED ISSUE #2179: RUC on type should affect virtual/override validation
```csharp
class Base {
	[RUC("RUC")]
	public virtual void Method() {}
}

[RUC("RUC")]
class Derived : Base {
	public override void Method() {}
}
```

Currently this warns about a mismatch on the base/override, but I think it shouldn't.

### PR DESCRIPTION
**Pull request by tlakollo**
Check also the declaring type when checking for attribute mismatch
Add tests
Fixes #2179

### CODE REVIEW

File: src/linker/Linker.Steps/ValidateVirtualMethodAnnotationsStep.cs [hunk 1]
```diff
@@ -41,8 +41,8 @@ protected override void Process ()
 		void ValidateMethodRequiresUnreferencedCodeAreSame (MethodDefinition method, MethodDefinition baseMethod)
 		{
 			var annotations = Context.Annotations;
-			bool methodHasAttribute = annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (method);
-			if (methodHasAttribute != annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (baseMethod)) {
+			bool methodHasAttribute = annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (method) || method.DeclaringType is not null && annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (method.DeclaringType);
+			if (methodHasAttribute != annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (baseMethod) || baseMethod.DeclaringType is not null && annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (baseMethod.DeclaringType)) {
```
Comment (sbomer) [inline src/linker/Linker.Steps/ValidateVirtualMethodAnnotationsStep.cs:45]: We already have a few places that perform similar checks to ask if a method is RUC (either directly or due to the type annotations). Elsewhere we still call TryGetEffectiveRequiresUnreferencedCodeAttributeOnType to deal with nested types but I think we are getting rid of that. Could we introduce a helper that encapsulates this logic to make sure it's consistent everywhere we call it?

**Review by sbomer — COMMENTED** [review]
Thanks!

File: src/linker/Linker.Steps/ValidateVirtualMethodAnnotationsStep.cs (hunk 1 above)
Comment (tlakollo) [reply src/linker/Linker.Steps/ValidateVirtualMethodAnnotationsStep.cs]: Replaced with DoesMethodRequireUnreferencedCode method

**Review by sbomer — APPROVED** [review]
Thank you!

**Pull request merged**